[Term Entry] JavaScript Array.prototype.toSorted() - close #2650 (#2716)
Codecademy/docs: content/javascript/concepts/arrays/terms/toSorted/toSorted.md

---
Title: '.toSorted()'
Description: 'Takes an array and returns a new array with all the elements sorted in ascending order.'
Subjects:
  - 'Web Development'
  - 'Computer Science'
Tags:
  - 'Arrays'
  - 'Comparison'
  - 'Functions'
  - 'Methods'
  - 'Sort'
  - 'Unicode'
CatalogContent:
  - 'introduction-to-javascript'
  - 'paths/front-end-engineer-career-path'
---

The **`toSorted()`** method takes an array and returns a new array with the elements sorted in ascending order. It does not mutate the original array. All undefined elements are sorted to the end of the array.

It does this by converting array elements to [strings](https://www.codecademy.com/resources/docs/javascript/strings) and comparing them by [Unicode code point value](https://en.wikipedia.org/wiki/Code_point). 

## Syntax

#### Functionless
```pseudo
array.toSorted();
```

#### Arrow function

> 'a' is the first element for comparison and 'b' is the second.

Author: Updates made in latest commit.

```pseudo
array.toSorted((a, b) => { /* Code for a compare function. */ });
```

#### Compare function
'compareFn' specifies a function that defines the sort order.
```pseudo
function compareFn() {
  /* Code for a compare function. */
};

array.toSorted(compareFn);
```

#### Inline compare function
```pseudo
array.toSorted(function compareFn(a, b) { /* … */ });
``` 

> ## Examples

Author: Updates made in latest commit.

const numbers = [-3, 4, 1, 9, 3, 2, 1, 200, 2159.20, 959, undefined, ""];

let sportsSorted = sports.toSorted();
let numbersSorted = numbers.toSorted();

console.log('Sports: ' + sportsSorted);
console.log('Numbers: ' + numbersSorted);
```
Which outputs:

```shell
Sports: ,baseball,basketball,pickleball,tennis,
Numbers: ,-3,1,1,2,200,2159.2,3,4,9,959,
```

`sports` was sorted alphabetically but `numbers` was sorted by unicode value. Remember that `toSorted()` converts elements to strings and compares them using Unicode Code Point value.

Additionally, `undefined` values were sorted to the end and empty strings (Unicode value of 0) were sorted to the beginning for both arrays.

To sort by numerical values, we require a comparison function.

### Codebyte Example

The code below takes the `numbers` array and sorts it two ways - ascending then descending. It then saves the new arrays to two variables and logs those. 

```codebyte/js
const numbers = [-3, 4, 1, 9, 3, 2, 1, 200, 959, 2159.20];

let ascending = numbers.toSorted(function compareFn(a, b) {
  return a-b});

let descending = numbers.toSorted(function compareFn(a, b) {
  return b-a});

> console.log('Numbers Ascending: ' + ascending); // Outputs "Numbers Ascending: -3,1,1,2,3,4,9,200,959,2159.2"
> console.log('Numbers Descending: ' + descending); // Outputs "Numbers Descending: 2159.2,959,200,9,4,3,2,1,1,-3"

Reviewer: ```suggestion
console.log('Numbers Ascending: ' + ascending); 
console.log('Numbers Descending: ' + descending); 
```
There's no need to add comment about the output as the viewer can run these code blocks themselves

---
Title: '.toSorted()'
Description: 'Takes an array and returns a new array with all the elements sorted in ascending order.'
Subjects:
  - 'Web Development'
  - 'Computer Science'
Tags:
  - 'Arrays'
  - 'Comparison'
  - 'Functions'
  - 'Methods'
  - 'Sort'
  - 'Unicode'
CatalogContent:
  - 'introduction-to-javascript'
  - 'paths/front-end-engineer-career-path'
---

> The **`toSorted()`** method takes an array and returns a new array with the elements sorted in ascending order. It does not mutate the original array. All undefined elements are sorted to the end of the array.

Reviewer 2: ```suggestion
The **`.toSorted()`** method takes an array and returns a new array with the elements sorted in ascending order. It does not mutate the original array. All undefined elements are sorted to the end of the array.
```

Author: 🤦‍♂️ Good catch.

> It does this by converting array elements to [strings](https://www.codecademy.com/resources/docs/javascript/strings) and comparing them by [Unicode code point value](https://developer.mozilla.org/en-US/docs/Glossary/Code_point). 

Reviewer 2: ```suggestion
The sorting is done by converting array elements to [strings](https://www.codecademy.com/resources/docs/javascript/strings) and comparing them by Unicode code point value. 
```

## Syntax

> **Functionless**

Reviewer 2: I think it may be preferable to combine all of these in one code block and the text in bold can be added to the right of the syntax as a comment

Author: @caupolicandiaz I initially did that but with the code in pseudo and the few lines of additional explanation, it was a little hard to parse. Is that a requirement for you to approve?

Author: N/m I'll make it work.

Author: ✅ in [5a59c46](https://github.com/Codecademy/docs/pull/2716/commits/5a59c46fe45eb19a8b49792b070baee48386a5a4).

```pseudo
array.toSorted();
```

**Arrow function**
'a' is the first element for comparison and 'b' is the second.

```pseudo
array.toSorted((a, b) => { /* Code for a compare function. */ });
```

**Compare function**
'compareFn' specifies a function that defines the sort order.

```pseudo
function compareFn() {
  /* Code for a compare function. */
};

array.toSorted(compareFn);
```

**Inline compare function**
```pseudo
array.toSorted(function compareFn(a, b) { /* … */ });
``` 

## Example

> In the below example, there are two arrays - `numbers`, containing integers & floats (a number with a decimal place), and `sports`, containing strings.

Reviewer 2: ```suggestion
In the example below, the `.toSorted()` method is applied to the arrays `sports` and `numbers`.
```

```js
const sports = ["baseball", "basketball", "tennis", "pickleball", undefined, ""]
const numbers = [-3, 4, 1, 9, 3, 2, 1, 200, 2159.20, 959, undefined, ""];

let sportsSorted = sports.toSorted();
let numbersSorted = numbers.toSorted();

console.log('Sports: ' + sportsSorted);
console.log('Numbers: ' + numbersSorted);

> ```

Reviewer 2: add an empty line

Author: ✅ in [5a59c46](https://github.com/Codecademy/docs/pull/2716/commits/5a59c46fe45eb19a8b49792b070baee48386a5a4).

Which outputs:

```shell
Sports: ,baseball,basketball,pickleball,tennis,
Numbers: ,-3,1,1,2,200,2159.2,3,4,9,959,
```

`sports` was sorted alphabetically but `numbers` was sorted by unicode value. Remember that `toSorted()` converts elements to strings and compares them using Unicode Code Point value.

Additionally, `undefined` values were sorted to the end and empty strings (Unicode value of 0) were sorted to the beginning for both arrays.

To sort by numerical values, we require a comparison function.

> ## Codebyte Example

Reviewer 2: This function is not currently supported by the Codecademy infrastructure so the codebyte will throw an error. Please remove this example.

Author: Ugh you're totally right. I forgot to check what version of node it was running. 

Any opposition to moving that to the example section? I think most people hitting the article will want to sort numerically and the method doesn't behave intuitively for sorting numbers numerically.

Author: Fixed in #[5a59c46](https://github.com/Codecademy/docs/pull/2716/commits/5a59c46fe45eb19a8b49792b070baee48386a5a4).
